### ISSUE TITLE
Memory discussion thread

### ISSUE THREAD

**Issue by jrz97619761** [jrz97619761/test-model-thing#20]
I think I'll put a central thread here for discussing architectural changes related to memory and training so everything is in one place.

If there is a bug you should report it separately in another issue but please keep the discussion and planning over here. And if you want to make a pull request make it separately as well.

Thanks

**Comment by reyvvon** [comment]
I still approve of a Discord but if you think this'll do fine it is your call

**Comment by jrz97619761** [comment]
I'll try to set one up at some point but it'll take some time because I just don't have the time to do it right now. I'm just mainly concerned about the risk of someone sending spam or otherwise disrupting a discord server, if people who don't know what they're doing decide to mess around in there.

Also I think doing it here means direct integration with code, PRs, etc and also it's more organized.

But I'll make a decision later, who knows.

**Comment by SimonVitzethum** [comment]
Great idea to keep this in one place. Here's a summary of what the CUDA fork ([SimonVitzethum/test-model-thing-CUDA](https://github.com/SimonVitzethum/test-model-thing-CUDA)) changed relative to the original, and why. Details are in [CHANGES.md](https://github.com/SimonVitzethum/test-model-thing-CUDA/blob/main/CHANGES.md) and `tmt-cuda/README.md`.

## Architecture

| Area | Original | CUDA fork | Why |
|---|---|---|---|
| Layer wiring | every layer reads the byte embedding | each layer reads the residual stream of the previous one | hierarchical features; higher layers can build on lower ones |
| Recurrence | `state = decay·state + enc` | `a = sigmoid(decay + gate·x)`, `s = a·s + (1−a)·x` | the additive state grows without bound; the normalized form stays bounded, and the input-dependent gate lets a channel decide what to write |
| Timescales | one decay per channel, half-life ~1 byte at init | half-lives spread from 2 to 512 bytes (configurable up to 100k) | channels for short and long memory from the start |
| Feedforward | one linear layer per layer | dense or Mixture-of-Experts (top-k), with load-balancing and z-loss | more capacity per compute; the losses prevent router collapse (an early run lost 114/256 experts without them) |
| Long range | recurrent state only | recurrent state; **optional** MLA attention cache (off by default) | kept optional; the default model relies on the recurrent state only |
| Losses | JEPA latent + CE + variance + stop | next-byte CE by default; the others are optional | simpler, and the extra terms can be tested as ablations |
| Memory across runs | state saved into checkpoints and carried into chat | state saved only for exact training resume; eval/sampling start fresh | reproducible runs, no drift into inference |

## Training and gradients through time

- **Batched windows instead of byte-by-byte:** each step processes B streams × T bytes (e.g. 16 × 128) with one update. *Why:* per-byte updates mean one GPU sync per byte, which dominates the runtime.
- **Truncated BPTT within a window:** exact gradients over T bytes for *all* weights, including the carried state at the window edge. *Why:* the original traces covered only embedding and decay; every other weight got a one-step gradient.
- **Hybrid traces (`traces=1`, the "middle way"):** decay, gate and embedding carry exact traces across window boundaries, combined with the window's gradient at the boundary. This gives them an unbounded gradient horizon again. With `seqlen=1`, it reduces exactly to the original trace rule (tested). *Why:* to keep the trace philosophy (no maximum gradient window for these parameters) while the rest trains efficiently.
- **Original trace rule (`origtrace=1`, branch `original-trace-rule`):** a port of `main.py`'s rule, including one embedding trace per layer.
- **Per-byte trace decay (`trace_decay`) and document resets (`docsep`):** the reset works by forcing `a_t = 0` at a separator byte, so state, gradient and traces are cut by the same recurrence.
- **Optimizer:** AdamW with warmup/cosine schedule, gradient clipping, and a guard that refuses updates with non-finite gradients.

**How close are the gradients to exact BPTT?** Cosine similarity over 4,096 bytes on a trained checkpoint (1 = exact):

| Parameters | 1-step, no traces | original rule (1 byte + traces) | TBPTT 128 | hybrid (128 + traces) |
|---|---|---|---|---|
| gate | 0.28 | 0.91 | 0.80 | 0.99 |
| decay | 0.54 | 0.69 | 0.97 | 0.97 |
| feedforward / experts | 0.60 | 0.60 | 0.95 | 0.95 |
| router | 0.45 | 0.45 | 0.95 | 0.95 |

Traces work very well where they exist. The remaining gap for a fully window-free, trace-only model is the feedforward weights, which also influence the future through the next layer's recurrence. Tracing them is possible but memory-heavy with MoE; that's the main open design question for "traces everywhere".

## Infrastructure

- **Pure CUDA C++, no Python in training:** cuBLAS matmuls on tensor cores, fused kernels (the recurrence runs in 3 kernel launches per window), persistent memory arenas.
- **Throughput:** ~80k bytes/s for a 34M model on an RTX 5070 laptop.
- **Checkpoints:** versioned, with checksums and exact resume of weights, optimizer, data position and stream state.
- **Tests without Python:** finite-difference gradient checks for every component, an FP64 reference for the attention paths, resume and corruption tests, and memcheck-clean.
- **Diagnostics:**
  - `gradcheck` compares windowed gradients with exact BPTT.
  - The training log shows the trace-vs-window gradient ratio and the state magnitude per half-life bucket.
  - Evaluation can reset the state every N bytes to measure how much context the model actually uses.

## Results so far (enwik8, 34M params, 8h on a laptop GPU)

- **Held-out BPC 1.82** with the hybrid (an earlier run reached 2.50, but it had a MoE gradient bug that kept the experts untrained; fixed and documented).
- Resetting the state every 128 bytes costs 0.52 BPC (2.35 vs 1.82), and the model still gains between 1k and 4k bytes, so the recurrent state carries context far beyond the gradient window.
- Samples are now mostly real English words and plausible grammar, but not yet meaningful content.
- **Caveat:** the MoE fix, traces and longer half-lives changed at the same time. A `traces=0` control run is still needed to attribute the gain.

## Knowledge-graph memory (branch `knowledge-graph`)

This builds on @ProfessionalBeginner's idea. Facts from Wikidata are read through cross-attention into the model's state. On entities never seen in training:

- **Stage 1** (the right facts are provided): 90.3% exact answers, and 4% with another entity's facts. So the model uses the facts' content, and the facts live in the graph rather than in the weights.
- **Stage 2** (the model finds the entity itself, from its own state, among 18k nodes): 99.25% retrieval accuracy, and 79.7% end-to-end vs 79.9% with the correct facts provided.

If we want everything in the weights instead, a learned memory layer (product-key memory) could reuse the same machinery.

## Open questions for this thread

1. **Traces everywhere:** how to trace the feedforward/expert weights affordably (shared across streams, low-rank, diagonal approximations), and how far a window-free model gets.
2. **Parallel training (#9) and peer-to-peer (#8):** weight averaging every n bytes vs synchronous batched gradients; one mechanism for single-GPU, multi-GPU and across machines.
3. **Simplified losses:** is next-byte CE alone enough, or do the latent/stop terms help at larger scale?
4. **Memory in the weights vs an external graph:** a learned memory layer would keep everything internal, but facts could no longer be edited without training.

**Comment by jrz97619761** [comment]
Can you try to port some of these changes to MLX rather than CUDA? I think it would be helpful to support Apple silicon as well as Linux CUDA/CPU.

Also, different question, but would a dense layer be better or an MoE layer in your opinion? Tell me what you think and perhaps see the difference in testing.

**Comment by SimonVitzethum** [comment]
*MLX port:*  yes, I can try that later. I have an M2 MacBook, so I can develop and test on Apple silicon directly. I'd port the algorithmic changes rather than the CUDA kernels:

- windowed training over several streams at once (the main speed win)
- the hybrid traces (exact traces for decay/gate/embedding across windows, which reduces to your original rule with 1-byte windows)
- the normalized gated recurrence with spread half-lives
- CE-only loss by default

MLX's `mx.compile` and `mx.fast.metal_kernel` should cover the performance-critical recurrence scan. Since MLX now also has CUDA and CPU backends on Linux, one MLX codebase could run everywhere, with the CUDA version as a fast reference to check results against.

**Dense vs. MoE:** I'd lean towards MoE for this project, with some caveats.

*Why MoE fits here:*

- **More capacity at the same compute per byte.** With 8 experts and top-2 routing, the model stores 8 feedforward blocks but only runs 2 per byte. A byte-level model has to learn a lot of surface detail (spelling, markup, numbers, several languages), and extra parameters help there while the per-byte cost barely changes.
- **Specialization.** Experts can split by kind of input: prose vs. wiki markup, digits, names, other scripts. The router learns that split on its own.
- **It fits the recurrent design.** The recurrent state carries the sequence information and the feedforward layer does the per-byte processing. That's exactly where MoE adds capacity without touching the memory mechanism.
- **Scaling.** If this ever grows, MoE is the main way to get large-model quality at small-model compute. Most recent efficient models use it.

*The costs:*

- **More memory:** all experts plus their optimizer state have to be stored, about 8× the feedforward parameters.
- **Routing adds failure modes:** without load balancing, an early run lost 114 of 256 experts, and I also had a gradient bug in the expert backward (now fixed and covered by a test). Dense is simpler and harder to break.
- **Harder for "traces everywhere":** tracing the feedforward weights costs memory per expert, so it's heavier with MoE than with a dense layer.
- **Small batches hurt MoE efficiency,** since each expert only sees part of the batch. At 16×128 bytes per step that's fine; at 1 byte per step it isn't.

*Testing:* I haven't done a clean comparison yet. My latest 8h run (MoE, 8 experts, top-2) reached 1.82 held-out BPC on enwik8, but there's no dense control. I'll run one: dense vs. MoE at **equal wall-clock time** and at **equal active compute**, same data and seed. That tells us whether the extra capacity pays for its overhead at this size.

**Comment by SimonVitzethum** [comment]
I can also try an experimental ROCm port, but I wouldn't be able to test it myself.

**Comment by reyvvon** [comment]
@jrz97619761 I can help with the setup and moderation don't worry!

**Comment by SimonVitzethum** [comment]
https://github.com/SimonVitzethum/test-model-thing-CUDA/tree/mlx

**Comment by SimonVitzethum** [comment]
Should I make a pull request for MLX and CUDA?

**Comment by jrz97619761** [comment]
> Should I make a pull request for MLX and CUDA?

hmm, I think we'll keep upstream MLX only for now. feel free to open a set of smaller PRs since I prefer that over a single large one

**Comment by LifeSupport0** [comment]
Copied from the now-closed #25 

As per Issue #3 , the model has only 1 parameter associated with controlling each latent space dimension's persistence: an exponential decay.

Each token step, the model performs a Hadamard Product on the latent space dimension (in layer order) with a vector of the learned exponential decay weights. In cases as observed in #3, those exponential decays grow to the point where individual tokens decay so much information from dimensions that meaningful associations cannot be made with those dimensions, unless the model non-trivially constructs an equilibrium with its other weights and biases.

To solve this, I suggest adding 6 parameters to each dimension of latent space, making a total of 7.

```
where:
x_t = latent state input vector at time t
W_w, W_b = Write Gate layer weights and biases
W_g = Write Gate vector
E_w, E_b = Erase Gate layer weights and biases
E_g = Erase Gate Vector
R_w, R_b = Read Gate layer weights and biases
R_g = Read Gate Vector
D = latent state decay vector
S_t = layer latent state at time t
S_{t+1} = layer latent state at time t+1
y_t = latent state output vector at time t

W_g = sigmoid(W_w ⦿ x_t + W_b)
E_g = sigmoid(E_w ⦿ x_t + E_b)
S_{t+1} = ((1-E_g) ⦿ S_t + W_g ⦿ x_t) ⦿ D
R_g = sigmoid(R_w ⦿ x_t + R_b)

y_t = S_{t+1} ⦿ R_g
```

Why each gate/parameter:
- The Write Gate selectively protects stored information from mutation
- The Erase Gate can quickly and selectively remove information without relying on very high exponential decay
- The Read Gate can selectively prevent stored latent state from appearing in the inputs to future layers

These changes would ONLY apply to the portions of the model which contain persisted latent state dimensions, meaning that for a 512 latent-dimension RNN, the 512 parameters used for the memory architecture would grow to 3584 (0.07% of a 4.5M param model).

**Comment by jrz97619761** [comment]
@LifeSupport0 Thank you, that has been on my todo list for a while (many people have suggested write/read/erase gates). I will implement it asap.

**Comment by jrz97619761** [comment]
Also, @LifeSupport0, how should I implement RTRL traces for this? I'm a bit stuck.

I would presume the code looks something like this (ignoring trace logic):

```python
class Layer(nn.Module):
    def __init__(self, dim: int, spread: int):
        super().__init__()

        # you could also just write: self.gates = [nn.Linear(dim, dim) for _ in range(3)] and call them with self.gates[n](x)
        self.write = nn.Linear(dim, dim)
        self.erase = nn.Linear(dim, dim)
        self.read = nn.Linear(dim, dim)

        # is there a cleaner way to do the decay spread logic?
        halflives = mx.exp(mx.linspace(0.0, math.log(float(spread)), dim))
        retention = mx.exp(-math.log(2.0) / halflives)
        self.decay = mx.log(retention) - mx.log1p(-retention)

    def __call__(self, x: mx.array):
        # or, once again: gated = [mx.sigmoid(layer(x)) for layer in self.gates] and access with gated[n]
        write = mx.sigmoid(self.write(x))
        erase = mx.sigmoid(self.erase(x))
        read = mx.sigmoid(self.read(x))

        decay = mx.sigmoid(self.decay)
        state = ((1 - erase) * state + write * state) * decay
        output = state * read

        return output, state, decay
```

**Comment by LifeSupport0** [comment]
> Also, @LifeSupport0, how should I implement RTRL traces for this? I'm a bit stuck.

I use tanh in my own layers, but that's mostly just a personal choice. It's a function that's differentiable and in the range [0, 1] that matters here.
This'll be a bit different since I'm running it on CUDA, but this should be roughly what you need
```
class Layer(nn.Module):
    def __init__(self, dim, spread):
        super().__init__()
        with torch.no_grad():
            halflives = torch.exp(torch.linspace(0.0, math.log(float(spread)), dim))
            retention = torch.exp(-math.log(2.0) / halflives)
            self.decay = nn.Parameter(torch.log(retention))
        self.norm = nn.LayerNorm(dim)
        self.weights = nn.Linear(dim, dim, bias=False)
        self.silu = nn.SiLU()
        self.gate_w = nn.Linear(dim, dim)
        self.gate_e = nn.Linear(dim, dim)
        self.gate_r = nn.Linear(dim, dim)
        with torch.no_grad():
            for g in (self.gate_w, self.gate_e, self.gate_r):
                g.weight.zero_()
                g.bias.zero_()

    def forward(self, enc, x, state):
        e = 1 + torch.tanh(self.gate_e(enc)) # erase gate
        decay = torch.exp(self.decay) * e
        w = 1 + torch.tanh(self.gate_w(enc)) # write gate
        state = decay * state + w * enc # store latent state
        n = self.norm(state)
        r = 1 + torch.tanh(self.gate_r(n)) # read gate
        out = x + self.silu(self.weights(n))
        return out * r, state, decay
```

**Comment by jrz97619761** [comment]
I meant, not the layer calculation, but the following section in the current code (sorry for being unclear):

```python
        c_range = (mx.arange(256) == c)[:, None].astype(mx.float32)
        for i, layer in enumerate(self.blocks):
            dlds = dlds_s[i]

            decay_embedtrace = layer.embedtrace * decays[i]
            embedtrace = decay_embedtrace + c_range
            grads["encoder"]["embed"]["weight"] += dlds * decay_embedtrace
            
            decaytrace = (decays[i] * layer.decaytrace) + (decays[i] * (1.0 - decays[i]) * layer.states)
            grads["blocks"][i]["decay"] = dlds * decaytrace

            layer.states = mx.stop_gradient(states[i])

            layer.decaytrace = mx.stop_gradient(decaytrace)
            layer.embedtrace = mx.stop_gradient(embedtrace)

        mx.eval(*[
            value
            for layer in self.blocks
            for value in (layer.states, layer.decaytrace, layer.embedtrace)
        ])
            
        self.updategrads(grads)
```

This needs to be changed to adapt to the r/w/e gates if we are going to use traces to compute them.

**Comment by LifeSupport0** [comment]
Generally, when you're writing a Write and Erase gate, the easiest way to do it is by just taking 1-erase and using it as write (since generally when you're erasing something you have something else to put in there). I did the approach of keeping them as separate independent operations in my own code, but again, personal taste thing. If it's better, the model will descend the gradient where those two parameters are equal (or close to equal)

The traces for the erase, write, and read gates should just be the derivative of the movement of those vectors (which is what the gradients are, the change in state between two points conceptualized as "time"). The formal math is a bunch of partial derivatives I don't have the understanding of GitHub Rich Text Editor to express right now.

The derivative should look something like:
```
erase_trace = erase * layer.decaytrace + erase * (1.0 - erase) * old_state
grads[...] = dlds * erase_trace
# tanh x is just bounding layer input to interval (0, 1)
write_trace = erase * layer.writetrace + write * (1.0 - write) * tanh(layer_input)
grads[...] = dlds * write_trace
```

You want to initialize the gates to zero (so it behaves identically to there being no gates) and allow the model to learn into using them.

P.S. damn, I completely forgot the embedding trace messes around somewhat with the partial derivatives.
Adding these gates is going to be a lot more involved than changing just that code.

**Comment by jrz97619761** [comment]
No need for the formal math. I know what derivatives are lol, no need 2 explain

I was referring to how the embedding trace interacts as you said, otherwise it would be not too difficult to implement as you wrote in your example there.

(sorry for the rushed comment, in the middle of something right now.)
